---
title: "Impute Covariate Data in RNA-sequencing Studies"
author:
- name: Brennan H Baker
  affiliation: University of Washington
  email: brennanhilton@gmail.com
package: RNAseqCovarImpute
output:
  BiocStyle::html_document
vignette: |
  %\VignetteIndexEntry{Impute Covariate Data in RNA-sequencing Studies}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

# Background

Missing data is a common problem in observational studies, as modeling techniques such as linear regression cannot be fit to data with missing points. Missing data is frequently dealt with using complete case analyses in which any individuals with missing data are dropped from the study. Dropping participants can reduce statistical power and, in some cases, result in biased model estimates. A common technique to address these problems is to replace or 'impute' missing data points with substituted values. Typically, for a given covariate, missing data points are imputed using a prediction model including other relevant covariates and the outcome of interest as independent variables. Rather than replacing a missing data point with one predicted value, multiple imputation generates several predictions, thereby allowing for uncertainty about the missing data. In a typical multiple imputation procedure: 1) N imputed data sets are created, 2) each data set is analyzed separately (e.g., using linear regression), and 3) estimates and standard errors across the N analyses are pooled using Rubin's rules. 

Research has shown that outcome information should be included in the predictive models to avoid biased estimates. This can be problematic in 'omics' studies with high dimensional outcome data. Fitting an imputation model where the number of independent variables is far greater than the number of individuals in the study is generally not possible. For instance, in RNA sequencing studies with upwards of 20,000 genes, far more than 20,000 participants would be needed in order to apply a standard multiple imputation procedure. Missing values can be imputed separately for each outcome variable, but that would require making upwards of 20,000 multiply imputed data sets (one set for each gene outcome). This computational inefficiency can be addressed by binning genes into smaller groups (e.g., bins with 1 gene per every 10 individual in the study). 

# RNAseqCovarImpute Demonstration

This package contains functions for imputing covariate data in RNA sequencing studies and conducting differential expression analysis using the limma-voom pipeline. The procedure is outlined below:

1: Bin the genes into smaller groups (around 1 gene per 10 individuals in the study). For example, with 880 individuals we make bins of 88 genes.

2: Use the first bin of genes plus all user-specified covariates in the predictor matrix to make N imputed data sets (using ‘mice’ R package).

3: Run RNA-seq analysis on only the genes in this particular bin. The analysis is run on all N imputed data sets separately. The coefficients and standard errors are then combined with Rubin's rules and P-values computed for this particular bin of genes.

4: Repeat steps 2 and 3 for all other gene bins.

5: Combine results for all gene bins and compute FDR adjusted P-values.

## Generate random data with missing covariate data

Normally you would have your own covariate and RNA-sequencing data. We generate random data below for the purpose of this demonstration.
```{r, message=FALSE, warning=FALSE, eval = requireNamespace(c("stringr", "tidyr"))}
library(dplyr)
library(stringr)
library(tidyr)
library(edgeR)
library(mice)
library(RNAseqCovarImpute)
# Generate random covariate data
set.seed(2023)
x1 <- rnorm(n = 500)
y1 <- rnorm(n = 500)
z1 <- rnorm(n = 500)
a1 <- rbinom(n = 500, prob = .25, size = 1)
b1 <- rbinom(n = 500, prob = .5, size = 1)
data <- data.frame(x = x1, y = y1, z = z1, a = a1, b = b1)

# Generate random count data
nsamp <- nrow(data)
ngene <- 500
mat <- matrix(stats::rpois(n = nsamp * ngene, lambda = sample(1:500, ngene, replace = TRUE)),
    nrow = ngene,
    ncol = nsamp
)

# Make fake ENSEMBL gene numbers
annot <- tibble(number = seq_len(ngene), name1 = "ENS") %>%
    mutate(ENSEMBL = str_c(name1, number)) %>%
    dplyr::select(ENSEMBL)
rownames(mat) <- annot$ENSEMBL

# Make DGE list and set rownames to ENSEMBL gene numbers above
alldg_full <- DGEList(mat, genes = annot)
rownames(alldg_full) <- annot$ENSEMBL

# With real data, you would likely want to filter out low count genes and normalize to library size
# keep <- rowMeans(cpm(alldg_full, log = TRUE)) > 0
# alldg_full <- alldg_full[keep, , keep.lib.sizes = FALSE]
# alldg_full <- calcNormFactors(alldg_full)
```

Now we simulate missingness in the random data generated above using the mice ampute function. The below code simulates missingness in the data for all but the first variable (x), which we consider the main predictor of interest in this example. See the ampute documentation for more details. 
```{r, message=FALSE, warning=FALSE}
# First get all combos of 0 or 1 for the 4 other variables (y, z, a, and b)
pattern_vars <- expand.grid(c(0,1), c(0,1), c(0,1), c(0,1))
# Then add back a column for the predictor of interest, which is never amputed, so the first col =1 the whole way down
pattern2 <- matrix(1, nrow = nrow(pattern_vars), ncol = 1)
pattern1 <- cbind(pattern2, pattern_vars)
# Remove last row which is all 1s (all 1s means no missingness induced)
pattern <- pattern1[seq_len(15), ]

# Specify proportion of missing data and missingness mechanism for the ampute function
prop_miss <- 25
miss_mech <- "MAR"
result <- ampute(data = data, prop = prop_miss, mech = miss_mech, patterns = pattern)

# Extract the missing data
ampute_i <- result$amp
# Ampute function turns factors to numeric, so convert back to factors
ampute_i <- ampute_i %>% mutate(
    a = as.factor(a),
    b = as.factor(b)
)

# Calculate the new sample size if NAs are dropped as in a complete case analysis
sample_size <- ampute_i %>%
    drop_na() %>%
    nrow()

# As shown below, we have 24.2% missingness.
100 * (nsamp - sample_size) / nsamp
```


```{r, message=FALSE, warning=FALSE, echo=FALSE}
ampute_i %>%
    head(10) %>%
    knitr::kable(digits = 3, caption = "The first 10 rows of simulated covariate data with missingness")
```


## 1: Bin the genes into smaller groups
The get_gene_bin_intervals function returns a data frame of intervals that can be used to bin the DGE list. By default, number of bins and genes per bin are set so that each bin has approximately 1 gene per 10 individuals/samples in the data. The default gene bin size can be changed with the n argument.

```{r, message=FALSE, warning=FALSE}
intervals <- get_gene_bin_intervals(alldg_full, ampute_i, n = 10)
```

```{r, message=FALSE, warning=FALSE, echo=FALSE}
intervals %>%
    head(10) %>%
    knitr::kable(digits = 3, caption = "The first 10 gene bins. Start and end columns indicate row numbers for the beginning and end of each bin. Number indicates the number of genes in each bin.")
```

Our goal is to bin genes randomly, so we must randomize the order of the genes in our DGE list. Without this step, genes would be binned together based on their sequential order within the chosen gene annotation (e.g., ENSEMBL or ENTREZ). 

```{r, message=FALSE, warning=FALSE}
# Randomize the order of gene identifiers
annot <- annot[sample(seq_len(nrow(annot))), ]
# Match order of the genes in the counts table to the randomized order of genes in the annotation
mat <- mat[annot$ENSEMBL, ]
# Recreate a DGE list with randomized order of genes
alldg_full <- DGEList(mat, genes = annot)
```

## 2-4: Make imputed data sets for each bin of genes and conduct differential expression analysis

The impute_by_gene_bin function loops through a DGE list using the gene bin intervals from the "get_gene_bin_intervals" function. It returns a list of sets of n imputed datasets, one per gene bin. For instance, if n = 100 and intervals contains 200 gene bin intervals, output will be a list of 200 sets of 100 imputed datasets. Each of the 200 sets are imputed using only the genes in one gene bin. This same procedure can be run more quickly in parallel with the impute_by_gene_bin_parallel function, which uses the 'PSOCK' backend implemented via the doParallel package.
```{r, message=FALSE, warning=FALSE, results = 'hide'}
gene_bin_impute <- impute_by_gene_bin(ampute_i,
    intervals,
    alldg_full,
    n = 3
)
```

The limmavoom_imputed_data_list function loops through the imputed data list (output from "impute_by_gene_bin" function) and runs RNA-seq analysis with the limma-voom pipeline. Users specify the formula for the RNA-seq design matrix and the main predictor of interest for which log fold-changes will be estimated. This procedure can also be run in parallel with the limmavoom_imputed_data_list_parallel function. 
```{r, message=FALSE, warning=FALSE}
coef_se <- limmavoom_imputed_data_list(
    gene_intervals = intervals,
    DGE = alldg_full,
    imputed_data_list = gene_bin_impute,
    n = 3,
    voom_formula = "~x + y + z + a + b",
    predictor = "x"
)
```

## 5: Combine the results from each imputed dataset using Rubin's rules
The final step is to combine the results from each imputed dataset using Rubin's rules. The argument "model_results" is the output from the "limmavoom_imputed_data_list" function above. The result is a table with one row per gene. The table includes coefficients (e.g., logFC values) standard errors, degrees of freedom, t-statistics, P-Values, and adjusted P-values from the limma-voom pipeline. Both the raw and empirical Bayes moderated statistics are reported (see limma package for details). 
```{r, message=FALSE, warning=FALSE}
final_res <- combine_rubins(
    DGE = alldg_full,
    model_results = coef_se,
    voom_formula = "~x + y + z + a + b"
)
```


```{r, message=FALSE, warning=FALSE, echo=FALSE}
final_res %>%
    dplyr::select(ENSEMBL, coef_combined, combined_p_bayes, combined_p_adj_bayes) %>%
    head(10) %>%
    knitr::kable(digits = 3, caption = "The top 10 genes associated with predictor x sorted by lowest P-value")
```

# Downstream gene set / pathway analysis

The compatibility of pathway and gene set enrichment methods with multiple imputation depends on their inputs. The RNAseqCovarImpute multiple imputation method produces one final list of genes with their associated t-statistics, log fold changes, and P-values for differential expression. Thus, the method is compatible with gene set enrichment analyses that utilize gene rankings such as ora, or gene level statistics such as camera, and gage. However, RNAseqCovarImpute is not compatible with gene set enrichment analyses that require as input a gene expression matrix or data at the individual sample level, as the nature of multiple imputation requires the creation of multiple gene expression matrices across each imputed dataset.

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```


